Scripts/Helpers/ProxyUI.cs

```diff
@@ -0,0 +1,549 @@
+                        break;
+                    case VisibilityControlType.alphaProperty:
+                        this.RestartCoroutine(ref m_BodyVisibilityCoroutine, AnimateBodyAlphaVisibility(value));
+                        break;
+                    case VisibilityControlType.materialSwap:
+                        SwapBodyToHiddenMaterial(value);
+                        break;
+                }
+            }
+        }
+
+        void OnDestroy()
+        {
+            // Cleanup cloned materials
+            foreach (var affordanceDefinition in m_AffordanceMap.AffordanceDefinitions)
+            {
+                var visibilityDefinition = affordanceDefinition.visibilityDefinition;
+                var visibilityType = visibilityDefinition.visibilityType;
+                if (visibilityType == VisibilityControlType.colorProperty || visibilityType == VisibilityControlType.alphaProperty)
+                {
+                    var materialsAndAssociatedColors = visibilityDefinition.materialsAndAssociatedColors;
+                    if (materialsAndAssociatedColors == null)
+                        continue;
+
+                    foreach (var materialToAssociatedColors in visibilityDefinition.materialsAndAssociatedColors)
+                    {
+                        var material = materialToAssociatedColors.firstElement;
+                        if (material != null)
+                            ObjectUtils.Destroy(material);
+                    }
+                }
+            }
+
+            var bodyVisibilityDefinition = m_AffordanceMap.bodyVisibilityDefinition;
+            var bodyVisibilityType = bodyVisibilityDefinition.visibilityType;
+            if (bodyVisibilityType == VisibilityControlType.colorProperty || bodyVisibilityType == VisibilityControlType.alphaProperty)
+            {
+                foreach (var kvp in m_BodyMaterialOriginalColorMap)
+                {
+                    var material = kvp.Key;
+                    if (material != null)
+                        ObjectUtils.Destroy(material);
+                }
+            }
+            else if (bodyVisibilityType == VisibilityControlType.materialSwap)
+            {
+                foreach (var material in m_BodySwapOriginalMaterials)
+                {
+                    if (material != null)
+                        ObjectUtils.Destroy(material);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Setup this ProxyUI
+        /// </summary>
+        /// <param name="affordances">The ProxyHelper affordances that drive visual changes in the ProxyUI</param>
+        /// <param name="proxyOrigins">ProxyOrigins whose child renderers will not be controlled by the PRoxyUI</param>
+        public void Setup(Affordance[] affordances, List<Transform> proxyOrigins)
+        {
+            // Prevent multiple setups
+            if (m_ProxyUISetup)
```

Comment:
Comments on these if statements are redundant with error messages.

Also, I wonder if this is something we can/should do with an Editor utility (custom Editor or EditorWindow?), rather than a runtime validation. Thoughts, @AndrewTHEManeri?

Reply:
The problem we'll run into there is that this function is instantiating the affordance map and making extra entries for affordances that are not customized.  Caching this data (but changing it when the source changes) would still require a runtime check